core/drivers/riscv_zkr_rng.c

```diff
@@ -4,46 +4,66 @@
  */
 
 #include <crypto/crypto.h>
+#include <encoding.h>
 #include <kernel/panic.h>
 #include <riscv.h>
 #include <rng_support.h>
 #include <tee/tee_cryp_utl.h>
 
-#define OPST_BIST 0b00
-#define OPST_WAIT 0b01
-#define OPST_ES16 0b10
-#define OPST_DEAD 0b11
-
-TEE_Result hw_get_random_bytes(void *buf, size_t len)
+static inline bool __must_check seed_get_random_u16(uint16_t *val)
 {
-	uint8_t *ptr = buf;
-	uint32_t val = 0;
+	unsigned int retry_count = 0;
+	uint32_t seed = 0;
+	uint32_t opst = 0;
 
-	while (len > 0) {
+	do {
 		/*
-		 * The seed register must be accessed using CSR read-write
-		 * instructions. The write operation is ignored and serves
-		 * to indicate polling and flushing.
+		 * The seed register must be accessed using CSR
+		 * read-write instructions. The write operation
+		 * is ignored and serves to indicate polling and
+		 * flushing.
 		 */
-		val = swap_csr(CSR_SEED, val);
+		seed = swap_csr(CSR_SEED, 0);
+		opst = seed & SEED_OPST;
 
-		switch (val >> 30) {
-		case OPST_BIST:
-		case OPST_WAIT:
-			continue;
-		case OPST_ES16:
-			*ptr++ = val & 0xff;
-			len--;
-			if (len > 0) {
-				*ptr++ = val >> 8;
-				len--;
-			}
-			break;
-		case OPST_DEAD:
+		switch (opst) {
+		case SEED_OPST_ES16:
+			*val = seed & SEED_ENTROPY;
+			return true;
+		case SEED_OPST_DEAD:
 			/* Unrecoverable self-test error */
-			return TEE_ERROR_BAD_STATE;
+			return false;
+		case SEED_OPST_BIST:
+		case SEED_OPST_WAIT:
 		default:
-			break; /* can't happen */
+			retry_count++;
+			cpu_relax();
```

Comment:
What do you expect `cpu_relax()` to do?

Reply:
Wait for multiple CPU cycles; we may also use the `pause` instruction [1] to reduce power consumption.

[1] https://github.com/riscv/riscv-isa-manual/blob/98a92230a0d4fa7b5a6e297c32177f9177afd674/src/zihintpause.adoc#L4